mixnet/node/src/lib.rs

```diff
@@ -77,205 +76,259 @@ impl MixnetNode {
 
-        Self::forward(
-            pool,
-            max_retries,
-            retry_delay,
-            Body::new_final_payload(payload),
+        self.forward(
+            payload,
             NymNodeRoutingAddress::try_from_bytes(&destination_addr.as_bytes())?,
         )
         .await
     }
 
-    async fn forward(
-        pool: &ConnectionPool,
-        max_retries: usize,
-        retry_delay: Duration,
-        body: Body,
-        to: NymNodeRoutingAddress,
-    ) -> Result<()> {
+    async fn forward(&self, msg: Body, to: NymNodeRoutingAddress) -> Result<()> {
         let addr = SocketAddr::from(to);
-        let arc_socket = pool.get_or_init(&addr).await?;
-
-        if let Err(e) = {
-            let mut socket = arc_socket.lock().await;
-            body.write(&mut *socket).await
-        } {
-            tracing::error!("Failed to forward packet to {addr} with error: {e}. Retrying...");
-            return mixnet_protocol::retry_backoff(
-                addr,
-                max_retries,
-                retry_delay,
-                body,
-                arc_socket,
-            )
+
+        self.message_tx.send(Message::new(addr, msg))?;
+        Ok(())
+    }
+}
+
+struct PacketForwarder {
+    config: MixnetNodeConfig,
+    message_rx: mpsc::UnboundedReceiver<Message>,
+    message_tx: mpsc::UnboundedSender<Message>,
+    connections: HashMap<SocketAddr, TcpStream>,
+}
+
+impl PacketForwarder {
+    pub fn new(
+        message_tx: mpsc::UnboundedSender<Message>,
+        message_rx: mpsc::UnboundedReceiver<Message>,
+        config: MixnetNodeConfig,
+    ) -> Self {
+        Self {
+            message_tx,
+            message_rx,
+            connections: HashMap::with_capacity(config.connection_pool_size),
+            config,
+        }
+    }
+
+    pub async fn run(mut self) {
+        loop {
+            tokio::select! {
+                msg = self.message_rx.recv() => {
+                    if let Some(msg) = msg {
+                        self.handle_msg(msg).await;
+                    } else {
+                        // Channel closed, we should shutdown the message handler thread
+                        return;
+                    }
+                },
+                _ = tokio::signal::ctrl_c() => {
+                    tracing::info!("Shutting down message handler thread...");
+                    return;
+                }
+            }
+        }
+    }
+
+    async fn try_send(&mut self, target: SocketAddr, body: &Body) -> Result<()> {
+        if let std::collections::hash_map::Entry::Vacant(e) = self.connections.entry(target) {
+            match TcpStream::connect(target).await {
+                Ok(tcp) => {
+                    e.insert(tcp);
+                }
+                Err(e) => {
+                    tracing::error!("failed to connect to {}: {e}", target);
+                    return Err(MixnetNodeError::Protocol(e.into()));
+                }
```

Comment:
I'm still curious if you aim to retry on `connect` failures, or not. If you want to retry, this should be changed. If you don't want to retry, this code may be correct, but then we can discuss. I still think we should try our best to establish connections as long as the retry config allows.

Reply:
Ah, yes, this should be changed, we should retry on error.

Reply:
After reviewing the code, I found that the code actually handles the error in the caller function. The `handle_retry` fn will handle the connect failures.

```rust
    async fn send(&mut self, msg: Message) {
        if let Err(err) = self.try_send(msg.target, &msg.body).await {
            match err {
                MixnetNodeError::Protocol(ProtocolError::IO(e))
                    if e.kind() == std::io::ErrorKind::Unsupported =>
                {
                    tracing::error!("fail to send message to {}: {e}", msg.target);
                }
                _ => self.handle_retry(msg),
            }
        }
    }
```

Reply:
Ah, you're right. Sorry. I misread the `match` clause.